### DISCUSSION TITLE
zh/docs/rbac-with-conditions-api/

### DISCUSSION THREAD

**Discussion by giscus** [casbin/casbin#1352] [Docs comments]
# zh/docs/rbac-with-conditions-api/

A more user-friendly API for RBAC with conditions.

https://casbin.org/zh/docs/rbac-with-conditions-api/

**Comment by Deepak7985** [comment]
I want to create a role which has all the permission in my application -> admin
another role which has same access as above but not to delete any resource -> user
How can this be done ??

**Reply by hsluoyz** [reply]
@Deepak7985 Ask Casbin AI Assistant: https://ai.casbin.com/

**Reply by Deepak7985** [reply]
hey i tried , but its of no use

**Reply by Deepak7985** [reply]
I have an admin policy defined as

`p, admin, * , * `

this ideally should give all permission to admin if i am not wrong.

but when i'm trying to access some resource this is not allowing me to .
I have assined this role for a userId = 3

`g, 3, admin`

**Comment by epistemer** [comment]
I'm sorry but I don't understand where in the go code are the `startTime` and `endtime` passed ? how do they reach the policy.csv file from the SDK ?

**Reply by hsluoyz** [reply]
See the source code by yourself

**Reply by epistemer** [reply]
My question might've been a little bit vague, I'm not asking about the source code, I'm asking about the provided example above. The policy.csv file shows timestamps that are nowhere to be found in the go code, only the time matching function is provided.

**Reply by epistemer** [reply]
where did those timestamps come from, and how can I add my own ?

**Comment by AbhijithMR-403** [comment]
has = e.has_permission_for_user("alice", "read")

you needs to also mention the obj
like:
has = e.has_permission_for_user("alice", "data1", "read")

**Comment by xiaoLangtou** [comment]
Is the `hasPermissionForUser` method buggy here?

My current model is:

```
[matchers]
m = g(r.sub, p.sub) && keyMatch3(r.obj, p.obj) && r.act == p.act
```

Policy example: a route like `/abs/{id}` should match an actual request route `/abs/124`.

When I use `hasPermissionForUser`, the result is `false`, but when I directly use `enforce`, the result is `true`.

Logs:

```
true  test role permission enforce Admin [ '/user/detail/31', 'GET' ]
false test role permission hasPermissionForUser Admin [ '/user/detail/31', 'GET' ]
false test role permission hasPermissionForUser Admin [ '/user/detail/31', 'GET' ]
```

**Reply by hsluoyz** [reply]
@xiaoLangtou enforce and hasPermissionForUser works differently, this is expected. For your case, you should use enforce or batchEnforce
